pyface/data_view/abstract_data_model.py

```diff
@@ -0,0 +1,272 @@
+            Whether or not the row can ever have child rows.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_row_count(self, row):
+        """ How many child rows the row currently has.
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+
+        Returns
+        -------
+        row_count : non-negative int
+            The number of child rows that the row has.
+        """
+        raise NotImplementedError
+
+    # Data value methods
+
+    @abstractmethod
+    def get_value(self, row, column):
+        """ Return the Python value for the row and column.
+
+        The values for column headers are returned by calling this method with
+        row equal to [].  The values for row headers are returned by calling
+        this method with column equal to [].
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+        column : sequence of int
+            The indices of the column as a sequence of length 0 or 1.
+
+        Returns
+        -------
+        value : any
+            The value represented by the given row and column.
+        """
+        raise NotImplementedError
+
+    def can_set_value(self, row, column):
+        """ Whether the value in the indicated row and column can be set.
+
+        The default method assumes the data is read-only and always
+        returns False.
+
+        Whether or a column header can be set is returned by calling this
+        method with row equal to [].  Whether or a row header can be set
+        is returned by calling this method with column equal to [].
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+        column : sequence of int
+            The indices of the column as a sequence of length 0 or 1.
+
+        Returns
+        -------
+        can_set_value : bool
+            Whether or not the value can be set.
+        """
+        return False
+
+    def set_value(self, row, column, value):
+        """ Set the Python value for the row and column.
+
+        The default method assumes the data is read-only and always
+        returns False.
+
+        The values for column headers can be set by calling this method with
+        row equal to [].  The values for row headers can be set by calling
+        this method with column equal to [].
+
+        Parameters
+        ----------
+        row : sequence of int
+            The indices of the row as a sequence from root to leaf.
+        column : sequence of int
+            The indices of the column as a sequence of length 0 or 1.
+        value : any
+            The new value for the given row and column.
+
+        Returns
+        -------
+        success : bool
+            Whether or not the value was set successfully.
```

Comment:
The problem I was running into here is that we can have failure, and exceptions were being too expensive in early sample code that I wrote, so I have been avoiding them in favour of explicit signalling.  For setting exceptions may be OK, because other than large drag-and-drop operations, the limiting factor is the user's ability to physically change the code; and we can probably make a faster codepath for drops if needed.

I'll play around with this to see what it looks like.